### DISCUSSION TITLE
Help With the Function(): SparseRowVarStd

### DISCUSSION THREAD

**Discussion by zjianyouyou** [satijalab/seurat#5400] [Q&A]
I come across this question when exploring how FindVariableFeatures works.

When I come to the isolated part of the codes, I don't know what the  code do with the variance and mean. In other words, when I get the rowVariance, rowMean, and the variance.expected, how to get the variance.standardized? 
I have try my best to find the answer, but gaining nothing. I hope someone here can help me.

little part of the Function:
>>>
  if (selection.method == "vst") {
    if (clip.max == "auto") {
      clip.max <- sqrt(x = ncol(x = object))
    }
    object=a
    hvf.info <- data.frame(mean = rowMeans(x = object))
    hvf.info$variance <- SparseRowVar2(mat = object, mu = hvf.info$mean, 
                                       display_progress = verbose)
    hvf.info$variance.expected <- 0
    hvf.info$variance.standardized <- 0
    not.const <- hvf.info$variance > 0
    fit <- loess(formula = log10(x = variance) ~ log10(x = mean), 
                 data = hvf.info[not.const, ], span = loess.span)
    hvf.info$variance.expected[not.const] <- 10^fit$fitted

    hvf.info$variance.standardized <- SparseRowVarStd(mat = object, 
                                                      mu = hvf.info$mean, sd = sqrt(hvf.info$variance.expected), 
                                                      vmax = clip.max, display_progress = verbose)

    colnames(x = hvf.info) <- paste0("vst.", colnames(x = hvf.info))
  }

**Comment by jyang635** [comment]
Hi,

 I am currently exploring it and I might be able to answer your question. 
![formula](https://user-images.githubusercontent.com/50032125/165628080-6acb749a-ff9c-4b2b-bc36-965393cf10ea.png)

This is the formula for standardized value. The numerators are observed UMI and gene mean (rowMean). The denominator is the expected variance (variance.expected) from the fit.

Once you get the standardized matrix, you can calculate the rowVariance and get the standardized variance.

Hope this is helpful.
Junkai

**Comment by chonga2022** [comment]
Error in order(hvf.info$vst.variance.standardized, decreasing = TRUE) : 
  argument 1 is not a vector
